src/_macosx.m

```diff
@@ -637,24 +668,32 @@ static CGFloat _get_device_scale(CGContextRef cr)
     View* view;
     PyObject* size;
     int width, height;
-    PyObject* obj;
     FigureCanvas* canvas;
+    PyObject* icon_path;
 
     if (!self->window) {
         PyErr_SetString(PyExc_RuntimeError, "NSWindow* is NULL");
         return -1;
     }
+    static char* kwlist[3] = { "canvas", "icon_path", NULL };
+    icon_path = NULL;
+    if (!PyArg_ParseTupleAndKeywords(args, kwds,
+                "O!|O&", kwlist,
+                &FigureCanvasType, &canvas,
+                &PyUnicode_FSDecoder, &icon_path)) {
+        return -1;
+    }
+    if (set_icon(icon_path)) {
+        return -1;
+    }
```

Comment:
Won't this cause us to fail if the icon_path wasn't found or wasn't valid? I think that setting the icon should be more of a "nice-to-have", than a "must-have" feature. I think this is partially why I was thinking the `set_icon` call would be better up in the manager as an "extra" rather than in the init itself.

Reply:
Yes, it is. As for whether it's optional or not, the icon is a keyword argument with default None. By default, the icon is not set.

I think there are two separate issues: what happens on an error while setting an icon, and how/where an icon should be set.

For the first, I think an error in the Obj-C should result in a Python exception. I don't think setting an icon should silently fail. It would be impossible to debug without compiling against a debug Python, with debug dependencies, and running in a debugger. That's a hassle. If it makes sense to handle the error at a higher level, that can be done.

For the second, I am a bit more ambivalent. The real question is, "Should the icon be handled via a context manager?" In other words, should it be reset every time after Matplotlib finishes running? If the answer is no, then it doesn't much matter to me: make it a property, make it a setter method, put it in the constructor, it's just style semantics in my opinion and how you want to handle (or not) the Python exception raised. If the answer is yes, though, then this design is wrong and should change. That would be an API change, which would be a whole different can of worms...

Reply:
Ahh, yes, I missed that the default was None, thanks for clarifying! I agree with everything you said :) 1. Raise an error for sure, and 2. set it in the initializer like you have, or a direct call to set_icon() from Python right after initialization.

The reason I have a preference for the set_icon() coming from Python is to maintain similarity to the other backends where they call `setWindowIcon()` or `iconphoto()` after the window setup, and the manager init I think remains consistent then. But, as you mentioned we should see if others have a strong preference. I think we are down to the details at this point and overall it is on the right path 🚀